### PR TITLE
Fix: move sequence truncation into BatchConverter

### PR DESCRIPTION
**Pull request by Alexperiments**
Fixes Issue #157 

- I added the truncation logic in the `__call__` of `BatchConverter` and added a `truncate` parameter in its constructor.
- I checked that `BatchConverter` is accessed through the `get_batch_converter` method of `Alphabet`, therefore I added a `truncate` parameter to this method. 
- I accordingly modified all the calls to this method in the rest of the code.
- In the test scripts, I set `truncate=True` since it is the default behavior.

As a side note: I've just noticed that the commit I made to sync my fork is listed in my PR. This is not intended, but I don't know how to remove it directly. If the PR gets approved I can eventually drop it while rebasing.

### CODE REVIEW

File: esm/data.py [hunk 1]
```diff
@@ -97,13 +97,15 @@ def __init__(
         prepend_bos: bool = True,
         append_eos: bool = False,
         use_msa: bool = False,
+        truncate: bool = True,
```
Comment (tomsercu) [inline esm/data.py:100]: better not modify the Alphabet class, `truncate` is only relevant for `BatchConverter`

File: esm/data.py [hunk 2]
```diff
@@ -133,11 +135,11 @@ def get_tok(self, ind):
     def to_dict(self):
         return self.tok_to_idx.copy()
 
-    def get_batch_converter(self):
+    def get_batch_converter(self, truncate: bool):
```
Comment (tomsercu) [inline esm/data.py:138]: if we make this argument an `Optional[int]` with default of `None` named something like `truncation_seq_length` - we can pass in the max length at which to truncate. And no truncation by default.

File: esm/data.py [hunk 3]
```diff
@@ -255,14 +257,17 @@ class BatchConverter(object):
...
         self.alphabet = alphabet
+        self.truncate = truncate
 
     def __call__(self, raw_batch: Sequence[Tuple[str, str]]):
         # RoBERTa uses an eos token, while ESM-1 does not.
         batch_size = len(raw_batch)
         batch_labels, seq_str_list = zip(*raw_batch)
         seq_encoded_list = [self.alphabet.encode(seq_str) for seq_str in seq_str_list]
+        if self.truncate:
+            seq_encoded_list = [seq_str[:1022] for seq_str in seq_encoded_list]
```
Comment (tomsercu) [inline esm/data.py:270]: Don't hardcode this number, use `self. truncation_seq_length`

File: examples/variant-prediction/predict.py [hunk 4]
```diff
@@ -156,7 +156,7 @@ def main(args):
             model = model.cuda()
             print("Transferred model to GPU")
 
-        batch_converter = alphabet.get_batch_converter()
+        batch_converter = alphabet.get_batch_converter(args.truncate)
```
Comment (tomsercu) [inline examples/variant-prediction/predict.py:159]: This wasn't added to `parser`. But tbh I'd leave this script as-is; if you pass in a `--sequence` longer than allowed you'll run into all kinds of trouble like pseudo-perplexity (`compute_pppl`) is undefined, and if you have mutations beyond max_length truncation means they be just ignored

File: tests/test_alphabet.py [hunk 5]
```diff
@@ -6,7 +6,7 @@
 def _test_esm1b(alphabet):
     import torch
 
-    batch_converter = alphabet.get_batch_converter()
+    batch_converter = alphabet.get_batch_converter(truncate=True)
```
Comment (tomsercu) [inline tests/test_alphabet.py:9]: You can leave these tests as-is (without argument), but add a test specifically specifying `truncation_seq_length=10` for example, and making sure the output makes sense (proper truncation, padding and EOS behave as expected when there is trunction or no truncation)

File: tests/test_readme.py [hunk 6]
```diff
@@ -25,7 +25,7 @@ def test_readme_2():
 
     # Load ESM-1b model
     model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()
-    batch_converter = alphabet.get_batch_converter()
+    batch_converter = alphabet.get_batch_converter(truncate=True)
```
Comment (tomsercu) [inline tests/test_readme.py:28]: argument should be optional, no need to modify the readme

**Review by tomsercu — CHANGES_REQUESTED** [review]
Thanks so much for your contribution! The place where you do the truncation logic in the `__call__()` function seems good to me! But I'd want to turn on the behavior with an optional int, left some comments.

File: esm/data.py (hunk 3 above)
Comment (Alexperiments) [reply esm/data.py]: I thought it was only useful to enforce sequences shorter than 1024. 
I changed the bool into an optional int as you suggested.

File: tests/test_alphabet.py (hunk 5 above)
Comment (Alexperiments) [reply tests/test_alphabet.py]: I added a test to check the behavior of truncated sequences

**Comment by Alexperiments** [comment]
Thank you @tomsercu for your feedback. Tell me if there is something else that I need to change.

File: scripts/extract.py [hunk 7]
```diff
@@ -91,11 +91,6 @@ def main(args):
             if torch.cuda.is_available() and not args.nogpu:
                 toks = toks.to(device="cuda", non_blocking=True)
 
-            # The model is trained on truncated sequences and passing longer ones in at
-            # infernce will cause an error. See https://github.com/facebookresearch/esm/issues/21
-            if args.truncate:
```
Comment (tomsercu) [inline scripts/extract.py:96]: Now argparser and L76 will need to be changed to include `truncation_seq_length`, you can set a default value of 1022 in argparser.

**Review by tomsercu — APPROVED** [review]
one more thing to actually fix the `extract.py` script itself,
but after that it's good to go!

File: scripts/extract.py (hunk 7 above)
Comment (Alexperiments) [reply scripts/extract.py:96]: Should I remove the `--truncate` parser argument, since now the truncation is fully handled by `truncation_seq_length`?

Comment (tomsercu) [reply scripts/extract.py:96]: yes removing is good. But I'd want to preserve the default behavior of this script, ie add to the argparser the flag `--truncation-seq-length` and set its default to 1022

Comment (Alexperiments) [reply scripts/extract.py:96]: Done!

Comment (tomsercu) [reply scripts/extract.py:96]: Great!
